File: SuperLauncher/ui/glass_frame.py
from PyQt6.QtWidgets import QFrame
from PyQt6.QtGui import QPainter, QColor, QPen, QBrush
from PyQt6.QtCore import Qt, QRectF

class GlassFrame(QFrame):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setAttribute(Qt.WidgetAttribute.WA_TranslucentBackground)
        self._tint = QColor(25, 25, 35, 220)
        self._border_color = QColor(255, 255, 255, 25)
        self._radius = 16

    def set_tint(self, r, g, b, a=220):
        self._tint = QColor(r, g, b, a)
        self.update()

    def paintEvent(self, event):
        rect = self.rect()
        if rect.width() < 4 or rect.height() < 4:
            return
        p = QPainter(self)
        p.setRenderHint(QPainter.RenderHint.Antialiasing)
        r = QRectF(rect).adjusted(2, 2, -2, -2)
        p.setBrush(QBrush(self._tint))
        p.setPen(QPen(self._border_color, 1))
        p.drawRoundedRect(r, self._radius, self._radius)

File: SuperLauncherCpp/src/ui/widgets/GlassFrame.cpp
#include "GlassFrame.h"
#include <QPainterPath>
#include <QGraphicsBlurEffect>

GlassFrame::GlassFrame(QWidget *parent)
    : QFrame(parent)
{
    setAttribute(Qt::WA_TranslucentBackground);
}

void GlassFrame::setBlurIntensity(qreal radius) { m_blurRadius = radius; }
void GlassFrame::setTintColor(const QColor &c) { m_tintColor = c; }
void GlassFrame::setBorderRadius(int r) { m_borderRadius = r; }
void GlassFrame::setBorderColor(const QColor &c) { m_borderColor = c; }

void GlassFrame::paintEvent(QPaintEvent *)
{
    QPainter p(this);
    p.setRenderHint(QPainter::Antialiasing);
    p.setRenderHint(QPainter::SmoothPixmapTransform);

    QPainterPath path;
    path.addRoundedRect(rect().adjusted(2, 2, -2, -2), m_borderRadius, m_borderRadius);

    p.fillPath(path, m_tintColor);

    QPen pen(m_borderColor, 1);
    p.setPen(pen);
    p.drawPath(path);
}
